chore: improve husky and lint-staged checks

KelvinOm opened the pull request:
## Description

1. Update husky, prettier and lint-staged then move them to devDependencies
2. Configure husky and lint-staged
3. Impriove rules for the lint commands
4.  Fix errors of eslint and prettier. 
5. Remove redundant files

## Type of change

> Please delete options that are not relevant.
- Chore (housekeeping or task changes that don't impact user perception)


## How Has This Been Tested?
- Manual

## Checklist:
### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag


### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

KelvinOm commented on app/client/.eslintrc.json:
```diff
@@ -19,19 +19,20 @@
   },
   "rules": {
-    "@typescript-eslint/no-explicit-any": 0,
+    "@typescript-eslint/no-explicit-any": "off",
```
Quite a lot of falls according to these rules. It is necessary to clarify with QA whether they need these checks.

сс @SatishGandham

KelvinOm commented on app/client/cypress/.eslintrc.json:
```diff
@@ -4,6 +4,12 @@
   },
   "rules": {
-    "cypress/no-unnecessary-waiting": 0
+    "@typescript-eslint/no-array-constructor": "off",
```
Quite a lot of falls according to these rules. It is necessary to clarify with QA whether they need these checks.

сс @SatishGandham

SatishGandham commented on app/client/.lintstagedrc:
```diff
@@ -1,5 +1,8 @@
-	"src/**/*.tsx": ["npx eslint --fix", "npx prettier --write"],
-	"cypress/**/*.js": ["npx eslint --fix", "npx prettier --write"],
-	"src/**/*.ts": ["npx eslint --fix", "npx prettier --write"]
+	"src/**/*.{js,ts,tsx}": ["npx eslint --fix"],
```
Shouldn't just warn and let users fix the errors manually?

KelvinOm replied:
It is common practice to make corrections on a pre-commit hook since you are in control of what will be sent to git anyway.

SatishGandham commented on app/client/packages/storybook/package.json:
```diff
@@ -4,6 +4,12 @@
+  "scripts": {
+    "storybook": "start-storybook -p 6006",
+    "build": "build-storybook",
+    "lint:ci": "echo 'No lint specified'",
```
Whats this?

KelvinOm replied:
In order for the code in our packages to be consistent and correspond to the main code base, scripts `lint:packages` and `prettier:package` were added to package.json. In these scripts, we go recursively through all our packages using `yarn-workspaces-foreach` and if we find a package without the necessary script then there will be an error. This is necessary to ignore the error.

This is how it looks in the console
![Снимок экрана 2023-03-22 в 11 15 25](https://user-images.githubusercontent.com/11555074/226840379-dc858399-d687-4215-b5b9-68cc678724ea.png)

jsartisan commented on app/client/cypress/.eslintrc.json:
```diff
@@ -4,6 +4,12 @@
   "rules": {
-    "cypress/no-unnecessary-waiting": 0
+    "@typescript-eslint/no-array-constructor": "off",
+    "@typescript-eslint/no-namespace": "off",
```
Are we adding these rules based on some standards?

KelvinOm replied:
I added a TS check for CI and lint-staged in the cypress folder. Here I just disabled the checks of some rules, as this requires significant changes.

jsartisan commented on app/client/package.json:
```diff
@@ -302,12 +304,8 @@
-      "pre-commit": "lint-staged && git-secrets --scan --untracked && git-secrets --scan -r"
-    }
+    "workbox-webpack-plugin": "^6.5.3",
+    "yarn-workspaces-foreach": "^0.2.0"
```
Which package manager? I didn't understand your comment :|

KelvinOm replied:
We are currently working on updating to Yarn 3 and when this happens this package will not be needed as we will be able to use the command [yarn workspaces foreach](https://yarnpkg.com/cli/workspaces/foreach)

We are also discussing [pnpm](https://pnpm.io/) as an alternative and it will also be no need for this package.

jsartisan reviewed (commented):
LGTM! just 2 questions/comments.

SatishGandham reviewed (dismissed):
Looks good. Thanks for taking it up.
Please ensure the build passes successfully with the changes [no unused vars], also check if we need to make any changes to EE repo to fix the errors.

riodeuno commented:
Also, I really appreciate that we're improving the checks and making the codebase more consistent with these checks. Thank you!

KelvinOm commented:
@riodeuno I added a TS check for the cypress folder, we didn't check it before, and since there are quite a lot of errors according to some rules, I temporarily disabled them. What required minor changes I fixed. 

I plan to contact the QA team about this. [See comment.](https://github.com/appsmithorg/appsmith/pull/21679#discussion_r1144359459)

In general, the checks have become stricter and cover all file extensions that are in the project. Please look at [this](https://github.com/appsmithorg/appsmith/pull/21679/files#diff-b8d2936a034f3ef827722f0cd18b81d5522b816bfe559ce499e8841f2142aeaf) and [this](https://github.com/appsmithorg/appsmith/pull/21679/files#diff-c62c0492ee5c5065686e8e1b3f8055fe156893114c6d41776dce0d7de7fdb086).

riodeuno commented:
Shall we run the `/ok-to-test` command here, and complete the PR checks?

KelvinOm commented:
> Shall we run the `/ok-to-test` command here, and complete the PR checks?

The changes do not affect the functionality and all checks(except E2E) were successfully passed locally. I would like to merge this so as not to delay the update cycle. I think while the tests are going on, we can add something in the release and the cycle will have to be repeated again.

KelvinOm commented:
/ok-to-test sha=13c08c0
